KIPs/kip-81.md

```diff
@@ -0,0 +1,773 @@
+    function updateVoterAddress(address _addr) external onlyMultisigTx;
+}
+```
+
+#### Notifying changes to StakingTracker
+
+To notify balance and voter account changes to StakingTracker contract, CnStakingV2 contract shall call the StakingTracker whenever after every change. For example,
+
+```solidity
+abstract contract CnStakingV2 {
+    function depositLockupStakingAndInit() payable beforeInit() {
+        // ...
+        stakingTracker.refreshStake(address(this));
+    }
+    function stakeKlay() payable {
+        // ...
+        stakingTracker.refreshStake(address(this));
+    }
+    function withdrawLockupStaking(address _to, uint256 _value) {
+        // ...
+        _to.transfer(_value);
+        stakingTracker.refreshStake(address(this));
+    }
+    function withdrawApprovedStaking(uint256 _approvedWithdrawalId) {
+        // ...
+        _to.transfer(_value);
+        stakingTracker.refreshStake(address(this));
+    }
+    function updateVoterAddress(address _addr) {
+        // ...
+        voterAddress = _addr;
+        stakingTracker.refreshVoter(address(this));
+    }
+}
+
+interface IStakingTracker {
+    /// @dev Re-evaluate Tracker contents related to the staking contract
+    /// @param staking  The CnStaking contract address
+    function refreshStake(address staking) external;
+
+    /// @dev Update a GC's voter address from given address
+    /// @param staking  The CnStaking contract address
+    function refreshVoter(address staking) external;
+}
+```
+
+#### Version identifier
+
+To distinguish the CnStakingContract and CnStakingV2 contracts, V2 will have VERSION value 2.
+
+```solidity
+abstract contract CnStakingV2 {
+    uint256 constant public VERSION = 2;
+}
+```
+
+### StakingTracker
+
+StakingTracker collects voting related data of each GC.
+
+- Staking amounts and voting powers \
+Stored separately in `Tracker` structs. Each `Tracker` struct can be updated between `trackStart` and `trackEnd` blocks, but becomes immutable afterwards, essentially freezing the voting powers. A `Tracker` struct is first created in `createTracker()`, and updated by `refreshStake()`
+
+- Each GC’s voter accounts \
+Stored in global mappings, and can be updated any time. Voter account mappings are updated by `refreshVoter()`
+
+#### Contract interface
+
+```solidity
+abstract contract StakingTracker {
+    struct Tracker {
+        // Tracker is only updated if trackStart <= block.number < trackEnd.
+        uint256 trackStart;
+        uint256 trackEnd;
+
+        // List of eligible nodes and their staking addresses.
+        // Initialized by createTracker().
+        address[] nodeIds;
+        mapping(address => address) rewardToNodeId;
+        mapping(address => address) stakingToNodeId;
+
+        // Balances and voting powers.
+        // Initialized by createTracker() and updated by refreshStake().
+        mapping(address => uint256) stakingBalances;
+        mapping(address => uint256) nodeBalances;
```

Comment:
@blukat29 What are the differences between stakingBalances and nodeBalances?

Reply:
Seems like 
- stakingBalances; // stakingContract => its balance
- nodeBalances; // nodeId => total staking balance of a GC member

Reply:
@kjhman21 You are right. I've added short comments in acfd52e.